# ***Exploratory Data Analisys - Análisis sobre los hábitos de vida del ciudadano***
## ➡️ **Limpieza de datos**
------

### 🗒️**Bloque de contenidos:**

1. [Importación de módulos](#1-importación-de-módulos)
2. [Carga del dataset descodificado](#2-carga-del-dataset-descodificado) 
3. [Tareas de limpieza](#3-tareas-de-limpieza)  
4. [Ejecución de la limpieza](#4-ejecución-de-la-limpieza)
5. [Validaciones](#5-validaciones)  
    5.1. [Eliminación de la variable `genero`](#51-eliminación-de-la-variable-genero)  
    5.2. [Eliminación de las filas duplicadas y eliminación de filas con más de la mitad de valores nulos](#52-eliminación-de-las-filas-duplicadas-y-eliminación-de-filas-con-más-de-la-mitad-de-valores-nulos)  
    5.3. [Categorización de `sueño_h` creando otra variable](#53-categorización-de-sueño_h-creando-otra-variable)  
    5.4. [Transformación de las variables numéricas a tipo numérica](#54-transformación-de-las-variables-numéricas-a-tipo-numérica)  
    5.5. [Transformación de las variables numéricas a tipo numérica](#55-transformación-de-las-variables-numéricas-a-tipo-numérica)  
6. [Guardado de dataset limpio](#6-guardado-de-dataset-limpio)
7. [Próximos pasos](#7-próximos-pasos)

### **1. Importación de módulos**

In [1]:
# Ajusto la ruta
%cd ..

c:\Users\sorim\Documents\Data_Science\Evolve\repos\projects-evolve\EDA


In [2]:
from scripts.cleaning import clean_completed
import pandas as pd


### **2. Carga del dataset descodificado**

In [3]:
df = pd.read_csv('./data/data_rename.csv', low_memory = False) # Me salía un warning y me recomendó poner este argumento a False.
df.head()

ejercicio asma cancer genero       grupo_edad                ingresos  \
0        No   No     No    NaN  Age 80 or older                 Refused   
1        No   No    Yes    NaN  Age 80 or older    $25,000 to < $35,000   
2       Yes   No     No    NaN     Age 55 to 59  $150,000 to < $200,000   
3       Yes  Yes     No    NaN              NaN     Don’t know/Not sure   
4       Yes   No     No    NaN     Age 40 to 44    $25,000 to < $35,000   

                                    educacion         empleo   estado  \
0  Graduated from College or Technical School        Retired  Alabama   
1                       Graduated High School  Self-employed  Alabama   
2  Graduated from College or Technical School        Retired  Alabama   
3                       Graduated High School        Retired  Alabama   
4        Attended College or Technical School    A homemaker  Alabama   

             IMC sueño_h fumar                    alcohol salud_mental_malos  \
0            NaN     8.0    No  No drinks in past 30 days                NaN   
1     Overweight     6.0    No  No drinks in past 30 days                NaN   
2     Overweight     5.0    No  No drinks in past 30 days                3.0   
3  Normal Weight     7.0   Yes  No drinks in past 30 days                NaN   
4  Normal Weight     9.0    No            2 days per week                NaN   

  salud_general salud_fisica_mala diabetes enf_cardio  
0     Very good               NaN      Yes         No  
1     Excellent               NaN       No         No  
2     Very good               2.0       No         No  
3     Excellent               NaN       No         No  
4          Fair               2.0       No         No

### **3. Tareas de limpieza**

Retomamos las conclusiones que llegamos en el [notebook 02](./02_Primer_analisis_descriptivo.ipynb):


1. Se **eliminará la variable `genero`** por su alto contenido en nulos (99%).

2. Eliminar filas **duplicadas** manteniendo la primera ocurrencia.

3. Se **eliminarán las filas** con **más de la mitad de valores nulos** ya que ese registro no nos aporta información.

7. Se **categorizará la variable `sueño_h`** poniendo rangos de:
    - Menos de 4h.
    - Entre 4 y 7h.
    - Más de 7h

4. Se transformará el **tipo de las variables** en `salud_mental_malos` y `salud_fisica_mala` a **tipo numérica**.*

5. Se creará una **categoría `Unknow`** en aquellas variables categóricas que contengan nulos

6. En el resto de variables que contiene nulos, **no se transformarán** para no crear sesgos ya que la información que los encuestados dan es de vital importancia.



A continuación, procedemos a realizar las correspondientes transformaciones.

### **4. Ejecución de la limpieza**

In [4]:
cols_num, cols_cat, df_clean = clean_completed(df, col_drop='genero')

-> Columna "genero" eliminada ✅

******************************

	 --> Procediendo a eliminar las filas duplicadas y las filas con más de la mitad de nulos

Número de filas duplicadas sin tener en cuenta la primera ocurrencia: 2806
Número de registros en el DF con duplicados -> 445132

Número de registros en el DF sin duplicados -> 442326
Número de registros en el DF con thresh-> 442288

Diferencia entre sin thresh y con thresh -> 38

 Duplicados y filas eliminadas ✅

******************************

	 --> Procediendo a categorizar los valores de la coluna "sueño_h":


 Variable categorizada ✅

******************************

-> Columnas numéricas detectadas (a convertir): ['salud_mental_malos', 'salud_fisica_mala']

	 --> Procediendo a cambiar el tipo de variable a numérica:

 Variables cambiadas de tipo a numérica ✅

******************************

-> Columnas categóricas detectadas (para rellenar NaNs): ['ejercicio', 'asma', 'cancer', 'grupo_edad', 'ingresos', 'educacion', 'empleo', 

### **5. Validaciones**

Una vez realizada la limpieza, vamos a proceder a verificar todos los cambios. Por ello nos vamos a traer las tareas de limpieza y las validaremos de una en una

#### ***5.1. Eliminación de la variable `genero`***

In [5]:
df_clean.columns == 'genero'

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False])

#### ***5.2. Eliminación de las filas duplicadas y eliminación de filas con más de la mitad de valores nulos***

> Número de filas duplicadas sin tener en cuenta la primera ocurrencia: 2806  
> Número de registros en el DF con duplicados -> 445132  
> Número de registros en el DF sin publicados -> 442326  
> Número de registros en el DF con thresh-> 442288  
>Diferencia entre sin thresh y con thresh -> 38  

In [6]:
len(df)

445132

In [7]:
len(df_clean) + 38

442326

In [8]:
len(df_clean)

442288

#### ***5.3. Categorización de `sueño_h` creando otra variable***

In [14]:
df_clean['sueño_cat'].value_counts()

sueño_cat
Entre 4 y 7h    269545
Más de 7h       161343
Menos de 4h       5959
NULO              5441
Name: count, dtype: int64

#### ***5.4. Transformación de las variables numéricas a tipo numérica***

In [10]:
df_clean[cols_num].dtypes

salud_mental_malos    float64
salud_fisica_mala     float64
dtype: object

#### ***5.5. Transformación de las variables numéricas a tipo numérica***

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 445132 entries, 0 to 445131
Data columns (total 17 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   ejercicio           445130 non-null  object
 1   asma                445130 non-null  object
 2   cancer              445129 non-null  object
 3   grupo_edad          436053 non-null  object
 4   ingresos            432200 non-null  object
 5   educacion           442749 non-null  object
 6   empleo              438936 non-null  object
 7   estado              445132 non-null  object
 8   IMC                 396326 non-null  object
 9   sueño_h             445129 non-null  object
 10  fumar               413355 non-null  object
 11  alcohol             404369 non-null  object
 12  salud_mental_malos  179900 non-null  object
 13  salud_general       445129 non-null  object
 14  salud_fisica_mala   177308 non-null  object
 15  diabetes            445129 non-null  object
 16  en

In [12]:
df_clean['alcohol'].value_counts(dropna=False)

alcohol
No drinks in past 30 days    186926
2 days per week              143528
1 day per week                65580
NULO                          40451
Don’t know/Not sure            3445
Refused                        2358
Name: count, dtype: int64

### **6. Guardado de dataset limpio**

In [13]:
df_clean.to_csv('./data/data_clean.csv', index=False)

### **7. Próximos pasos**

Una vez que está listo nuestro dataset, en el siguiente cuaderno ([Notebook 04](04_Visualizaciones_y_analisis.ipynb)), procederemos al análisis entre las variables para observar si hay algún tipo de relación entre ellas.

----